### PR TITLE
Query Input refactoring

### PR DESCRIPTION
**Pull request by agutierrezrodriguez**
Related with #114

### CODE REVIEW

File: demo/demo.js [hunk 1]
```diff
@@ -175,4 +175,35 @@
...
+  });
+  var secondResultsMultiple = new doofinder.widgets.ScrollDisplay("#secondResultsMultiple", {
+    template: "{{#results}}<li>{{id}}</li>{{/results}}"
+  });
+
+  firstControllerMultiple.registerWidgets([multipleInputWidget, firstResultsMultiple]);
+  secondControllerMultiple.registerWidgets([multipleInputWidget, secondResultsMultiple]);
+
+
+  window.doofinderControllers.multipleInputsAndControllers = {first: firstControllerMultiple, second: secondControllerMultiple};
```
Comment (carlosescri) [inline demo/demo.js]: You could add a `try..catch` block and try to assign the input to a different widget. When you catch the error add a log in the console explaining what happened.

File: src/widgets/queryinput.coffee [hunk 2]
```diff
@@ -1,6 +1,8 @@
 extend = require "extend"
 Widget = require "./widget"
 Thing = require "../util/thing"
+Error = require "../util/errors"
```
Comment (carlosescri) [inline src/widgets/queryinput.coffee]: `Error > errors` try to use the same conventions as in the rest of the code

File: src/widgets/queryinput.coffee [hunk 3]
```diff
@@ -41,6 +46,13 @@ class QueryInput extends Widget
       controller = [controller]
     @controller = @controller.concat controller
 
+  setElement: (elements) ->
+    @element = ($ elements).map (elem) =>
+      if elem.doofinderQueryInput?
+        throw Error.error("(#{elem.id}) was registered in another QueryInput", elem)
```
Comment (carlosescri) [inline src/widgets/queryinput.coffee]: No parenthesis, please.

File: src/widgets/queryinput.coffee [hunk 4]
```diff
@@ -50,13 +62,15 @@ class QueryInput extends Widget
...
-            @trigger "df:input:submit", [@value]
+      @element.on "input", (event) =>
+        @currentElement = $ event.target
+        # @element.val @value # Synchronize text inputs
+        @__scheduleUpdate()
+
+      @element.filter(":not(textarea)").on "keydown", (event) =>
+        if event.keyCode? and event.keyCode is 13
+          @__scheduleUpdate 0, true
+          @trigger "df:input:submit", [@value]
```
Comment (carlosescri) [inline src/widgets/queryinput.coffee]: I think a test is required for this to ensure that the value sent is the correct (`@currentElement` is the input and not a textarea). It may not be easy.

**Comment by carlosescri** [comment]
In any case, it looks very good 👍 . Tomorrow I'll give it a try through the demo page.

File: src/widgets/queryinput.coffee (hunk 4 above)
Comment (agutierrezrodriguez) [reply src/widgets/queryinput.coffee]: Already have a test for this.

File: src/widgets/queryinput.coffee [hunk 5]
```diff
@@ -1,7 +1,7 @@
 extend = require "extend"
 Widget = require "./widget"
 Thing = require "../util/thing"
-Error = require "../util/errors"
+error = require "../util/errors"
```
Comment (carlosescri) [inline src/widgets/queryinput.coffee]: `error -> errors`

Comment (agutierrezrodriguez) [reply src/widgets/queryinput.coffee]: sorry :disappointed:

File: src/widgets/queryinput.coffee [hunk 6]
```diff
@@ -47,9 +47,9 @@ class QueryInput extends Widget
     @controller = @controller.concat controller
 
   setElement: (elements) ->
-    @element = ($ elements).map (elem) =>
+    @element = ($ elements).filter (elem) =>
```
Comment (carlosescri) [inline src/widgets/queryinput.coffee]: why this change? (just to know)

Comment (agutierrezrodriguez) [reply src/widgets/queryinput.coffee]: Because map not work as expected, and tests are failing because this.
